baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/DangerousCollectionCopyOfUsage.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import java.util.Map;
+import java.util.Set;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "DangerousCollectionCopyOfUsage",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = BugPattern.SeverityLevel.ERROR,
+        summary = "Disallow Java 10+ java.util.{Map,Set}.copyOf methods.")
+public final class DangerousCollectionCopyOfUsage extends BugChecker implements BugChecker.MethodInvocationTreeMatcher {
+    private static final String ERROR_MESSAGE = "Should not use java.util.{Map,Set}.copyOf methods, and should instead "
+            + "prefer Guava Immutable{Map,Set}.copyOf. Java's internal immutable collections are not hash based, they "
+            + "are linear lookup based. This means that instead of O(1) lookup time, they have O(n) lookup time. This "
```

Comment:
I'm reading through the `ImmutableCollections.MapN` class to try to try and understand this, and while the design does seem unusual, I'd be surprised if you get O(n) for lookups??

```
/**
     * An array-based Map implementation. There is a single array "table" that
     * contains keys and values interleaved: table[0] is kA, table[1] is vA,
     * table[2] is kB, table[3] is vB, etc. The table size must be even. It must
     * also be strictly larger than the size (the number of key-value pairs contained
     * in the map) so that at least one null key is always present.
```

They use linear probing, but because and javadoc describes how it uses a flat array as internal storage, but I think the key thing is that they use a load factor of 2, so you shouldn't have to iterate over the whole thing.

The only method that did look gross to me was `containsValue`, cos that is _definitely_ O(n) whereas in guava that's definitely O(1) amortised